Update client config invalid option errors to be more specific

juanvallejo opened the pull request:
This patch adds better error handling for cases where a global option (such as --context or --cluster) causes an invalid config to be returned.

``` release-note
release-note-none
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29236)

<!-- Reviewable:end -->

liggitt commented on pkg/kubectl/cmd/get.go:
```diff
@@ -218,6 +219,20 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
+		// flag is set, such flag's value can then be blamed
+		if clientcmd.IsConfigurationMissing(err) {
+			context := cmdutil.GetFlagString(cmd, "context")
+			cluster := cmdutil.GetFlagString(cmd, "cluster")
```
We should definitely not be looking up specific flags by name and erroring here. If we error, it should be when we attempt to apply the command line override flags at point of use

liggitt commented:
I'd like to see the following tests in test-cmd.sh:

1. invalid --user prints appropriate message
2. invalid --context prints appropriate message
3. invalid --cluster prints appropriate message
4. context in KUBECONFIG containing invalid cluster reference behaves as it does today
5. context in KUBECONFIG containing invalid user reference behaves as it does today
6. missing KUBECONFIG file behaves as it does today
7. `kubectl config set-context new-context-name --cluster=missing-cluster --user=missing-user --namespace=some-namespace` still allows specifying `--cluster` and `--user` values which do not exist

juanvallejo commented:
@liggitt 
> I'd like to see the following tests in test-cmd.sh:

Could you take a look please? https://github.com/kubernetes/kubernetes/pull/29236/commits/8e1095bd2265d2e5a7f8895634662f4de8973ea9

caesarxuchao commented:
ok to test

caesarxuchao commented:
@liggitt hope you don't mind me reassigning this PR to you since you have started reviewing it. Thanks.

juanvallejo commented:
@liggitt could you take a look please? I think the last test might have flaked

liggitt commented on hack/make-rules/test-cmd.sh:
```diff
@@ -1519,6 +1519,51 @@ __EOF__
+  # Command
+  output_message=$(! kubectl get pod --context="" --kubeconfig=missing 2>&1)
+  kube::test::if_has_string "${output_message}" "missing: no such file or directory"
+  # Post-condition: busybox0 & busybox1 PODs are retrieved, but because busybox2 is malformed, it should not show up  kube::test::if_has_string "${output_message}" "missing: no such file or directory"
```
comment is weird

liggitt commented:
Does the description need updating? I don't see an error type for an entirely missing configuration file.

Otherwise, I think this looks good, though I'd probably hold it until 1.4 cuts at this point.

@deads2k, how do you feel about treating a missing override cluster/auth/context as an error? `kubectl --cluster=foo` not telling you that `foo` doesn't exist seems off

deads2k commented:
> @deads2k, how do you feel about treating a missing override cluster/auth/context as an error?

I'm fine with that.

juanvallejo commented:
@liggitt Thanks for the feedback!

juanvallejo committed:
handle invalid client config option errors

This patch provides a more relevant error message when a client
configuration option is passed with an invalid or non-existent value.

`$ kubectl get pods --cluster="non-existent"`
```
error: No configuration file found, please login or point to an existing
file
```

`$ kubectl get pods --cluster="non-existent"`
```
error: cluster "non-existent" does not exist
```

juanvallejo commented:
@liggitt could you please take a look at https://github.com/kubernetes/kubernetes/pull/29236/commits/6adf162344be6d38f55fede3e016b33724918fa1 ? I added an "ignoreCurrentContext" to the ClientConfig overrides so that the kubectl integration test would pass. The test sets an override for context `test`, which does not already exist. This patch requires any overridden value to already exist (in order to return an appropriate error in the case that it does not), causing the integration test to fail every time.

liggitt commented:
can you link to the integration test that is failing? I would not expect specifying a current context that does not exist to work well... do we need to reexamine the test?

juanvallejo commented:
@liggitt 
Link to test: https://github.com/kubernetes/kubernetes/blob/master/test/integration/kubectl/kubectl_test.go
It fails [here](https://github.com/kubernetes/kubernetes/blob/master/test/integration/kubectl/kubectl_test.go#L63) with the error "failed to get validator: context "test" does not exist" after [this change](https://github.com/kubernetes/kubernetes/pull/29236/files#diff-093eaa515b227f1070ec0fda3f8e0137R335) "requires" the overridden context to exist.

juanvallejo commented:
@liggitt Perhaps a possible change could be to alter [this `else if`](https://github.com/kubernetes/kubernetes/pull/29236/files#diff-093eaa515b227f1070ec0fda3f8e0137R335) to be something like `} else if required && config.overrides.Context == nil {` ?

juanvallejo commented on pkg/client/unversioned/clientcmd/client_config.go:
```diff
@@ -273,66 +293,79 @@ func (config *DirectClientConfig) ConfirmUsable() error {
 	var mergedContext clientcmdapi.Context
 	if configContext, exists := contexts[contextName]; exists {
 		mergo.Merge(&mergedContext, configContext)
+	} else if required {
```
@liggitt 
> I would not expect specifying a current context that does not exist to work well... do we need to reexamine the test?

Hm, [the test does seem to pass a context](https://github.com/kubernetes/kubernetes/blob/master/test/integration/kubectl/kubectl_test.go#L56) to `clientcmd.ConfigOverrides`.
Perhaps a possible change could be to alter this else if to be something like } else if required && config.overrides.Context == nil { ?

liggitt replied:
I'm still not seeing why the test needs to send an overridden "current context" at all, if it's overriding the actualy context values

juanvallejo replied:
@liggitt 
> I'm still not seeing why the test needs to send an overridden "current context" at all, if it's overriding the actual context values

I went ahead and removed the `CurentContext` override from the test. I kept the new `IgnoreMissingContext` [flag as a way to tell client_config when to 'require' a context](https://github.com/kubernetes/kubernetes/pull/29236/commits/6adf162344be6d38f55fede3e016b33724918fa1#diff-093eaa515b227f1070ec0fda3f8e0137R297) that has been overridden (the kubectl integration test otherwise still fails without it). PTAL

liggitt commented on hack/make-rules/test-cmd.sh:
```diff
@@ -1541,6 +1541,46 @@ __EOF__
+  kube::test::if_has_string "${output_message}" 'auth info "missing-user" does not exist'
+
+  # test invalid config
+  kubectl config view | sed -E "s/apiVersion: .*/apiVersion: v-1/g" > /tmp/newconfig.yaml
```
`${TMPDIR:-/tmp}`, don't assume write permission to /tmp

liggitt commented on hack/make-rules/test-cmd.sh:
```diff
@@ -1541,6 +1541,46 @@ __EOF__
   kubectl delete namespace test-configmaps
 
   ####################
+  # ConfigMap Errors #
```
fix test title

liggitt commented on hack/make-rules/test-cmd.sh:
```diff
@@ -1541,6 +1541,46 @@ __EOF__
+  ####################
+
+  ### Test errors due to invalid config flags
+  # Pre-condition: no POD exists
```
fix comment

liggitt commented on hack/make-rules/test-cmd.sh:
```diff
@@ -1541,6 +1541,46 @@ __EOF__
+  # test invalid config
+  kubectl config view | sed -E "s/apiVersion: .*/apiVersion: v-1/g" > /tmp/newconfig.yaml
+  output_message=$(! "${KUBE_OUTPUT_HOSTBIN}/kubectl" get pods --context="" --user="" --kubeconfig=/tmp/newconfig.yaml 2>&1)
+  kube::test::if_has_string "${output_message}" "failed to get client"
```
huh, that's not the message I would have expected for an invalid file

juanvallejo replied:
Ah, the entire error message is: `failed to get client config: Error loading config file "/tmp/invalidconfig.yaml": no kind "Config" is registered for version "v-1"`

liggitt commented on hack/make-rules/test-cmd.sh:
```diff
@@ -1541,6 +1541,46 @@ __EOF__
+  output_message=$(! kubectl get pod --kubeconfig=missing-config 2>&1)
+  kube::test::if_has_string "${output_message}" 'no such file or directory'
+
+  output_message=$(! kubectl get pod --cluster="" --user="" --context="" --server=org 2>&1)
```
this test seems problematic, bogus DNS names can hang or resolve in ways you wouldn't expect in different envs

liggitt commented:
a few test cleanups, then squash and LGTM post-1.4

juanvallejo commented:
@liggitt Thanks for the feedback!

liggitt commented:
@k8s-bot test this

liggitt commented on test/integration/kubectl/kubectl_test.go:
```diff
@@ -56,6 +56,7 @@ func TestKubectlValidation(t *testing.T) {
 	cfg.CurrentContext = "test"
 	overrides := clientcmd.ConfigOverrides{
 		ClusterInfo: *cluster,
+		Context:     *ctx,
```
remind me why this is needed, if CurrentContext is already set to "test"?

juanvallejo replied:
Hm, I agree, it looks to not be needed, I've tested this locally and the kubectl integration test is passing without it. I've gone ahead and removed it

liggitt commented:
LGTM
